File: WdtcCore/src/main/java/org/wdt/wdtc/core/download/InstallGameVersion.java
package org.wdt.wdtc.core.download;

import lombok.Setter;
import org.apache.log4j.Logger;
import org.wdt.utils.io.FileUtils;
import org.wdt.wdtc.core.download.infterface.InstallTaskInterface;
import org.wdt.wdtc.core.download.infterface.TextInterface;
import org.wdt.wdtc.core.game.Launcher;
import org.wdt.wdtc.core.game.config.GameConfig;
import org.wdt.wdtc.core.utils.ModUtils;
import org.wdt.wdtc.core.utils.WdtcLogger;

import java.io.IOException;

@Setter
public class InstallGameVersion extends DownloadGameVersion {
  private static final Logger logmaker = WdtcLogger.getLogger(InstallGameVersion.class);
  private TextInterface setTextFieldText;

  public InstallGameVersion(Launcher launcher, boolean Install) {
    super(launcher, Install);
  }

  public InstallGameVersion(Launcher launcher) {
    this(launcher, false);
  }


  public void InstallGame() {
    try {
      long startTime = System.currentTimeMillis();
      if (FileUtils.isFileNotExists(launcher.getVersionConfigFile())) {
        GameConfig.writeConfigJson(launcher);
      } else {
        GameConfig.ckeckVersionInfo(launcher);
      }
      DownloadGameFile();
      InstallTaskInterface task = ModUtils.getModInstallTask(launcher);
      if (install && task != null) {
        task.beforInstallTask();
        task.writeVersionJsonPatches();
        task.overwriteVersionJson();
      }
      DownloadGameLibrary();
      String LibraryFinishTime = "游戏所需类库下载完成,耗时:" + (System.currentTimeMillis() - startTime) + "ms";
      logmaker.info(LibraryFinishTime);
      if (setTextFieldText != null) {
        setTextFieldText.setControl(LibraryFinishTime);
      }
      if (install && task != null) {
        task.afterDownloadTask();
      }
      DownloadResourceFile();
      String EndTime = "下载完成,耗时:" + (System.currentTimeMillis() - startTime) + "ms";
      logmaker.info(EndTime);
      if (setTextFieldText != null) {
        setTextFieldText.setControl(EndTime);
      }
    } catch (IOException e) {
      logmaker.error("Download Game Error,", e);
    }
  }
}

File: WdtcCore/src/main/kotlin/org/wdt/wdtc/core/download/InstallGameVersion.kt
package org.wdt.wdtc.core.download

import kotlinx.coroutines.coroutineScope
import org.wdt.wdtc.core.download.game.DownloadVersionGameFile
import org.wdt.wdtc.core.game.Version
import org.wdt.wdtc.core.game.VersionsList.Companion.changeListToFile
import org.wdt.wdtc.core.game.currentVersionsList
import org.wdt.wdtc.core.manger.writeConfigJson
import org.wdt.wdtc.core.utils.launch
import org.wdt.wdtc.core.utils.logmaker
import org.wdt.wdtc.core.utils.modInstallTask

// TODO Optimize download speed
class InstallGameVersion(
	private val version: Version, private val install: Boolean = false,
	private val printInfo: (suspend (String) -> Unit)? = null
) {
	
	private val download: DownloadVersionGameFile = DownloadVersionGameFile(version, install)
	
	
	suspend fun startInstallGame(): Unit = coroutineScope {
		val startTime = System.currentTimeMillis()
		val task = version.modInstallTask
		
		download.createGameDirectories()
		
		launch("Write and ckeck config file") {
			if (install) {
				version.writeConfigJson()
			}
		}
		
		launch("Add version to versions list") {
			currentVersionsList.changeListToFile {
				version.addToList()
			}
		}
		
		startDownloadGameFileTask()
		
		if (install) task?.run {
			beforInstallTask()
			writeVersionJsonPatches()
			overwriteVersionJson()
		}
		
		download.startDownloadLibraryFile()
		
		System.currentTimeMillis().minus(startTime).also {
			printInfo?.invoke("下载游戏所需类库完成,耗时${it}ms")
			logmaker.info("Download game runtime finish,take a period of: ${it}ms")
		}
		
		if (install) task?.afterDownloadTask()
		
		download.startDownloadAssetsFiles()
		
		System.currentTimeMillis().minus(startTime).also {
			printInfo?.invoke("游戏下载完成,耗时${it}ms")
			logmaker.info("Download game finish,take a period of: ${it}ms")
		}
	}
	
	private suspend fun startDownloadGameFileTask() {
		coroutineScope {
			download.run {
				startDownloadGameVersionJson()
				launch("Download assets list json") { startDownloadGameAssetsListJson() }
				launch("Download version jar") { startDownloadVersionJar() }
			}
		}
	}
}

